File: epi_judge_java/epi/NQueens.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiTestComparator;
import epi.test_framework.GenericTest;
import epi.test_framework.LexicographicalListComparator;

import java.util.ArrayList;
import java.util.List;

public class NQueens {

    @EpiTest(testDataFile = "n_queens.tsv")
    public static List<List<Integer>> nQueens(int n) {
        List<List<Integer>> ans = new ArrayList<>();
        if (n == 1) {
            ans.add(List.of(0));
            return ans;
        }
        if (n <= 3) return ans;
        List<Integer> currPosition = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            currPosition.add(-1);
        }
        recur(0, n, currPosition, ans);
        return ans;
    }

    private static void recur(int row, int n, List<Integer> currPosition, List<List<Integer>> ans) {
        if (row == n) {
            ans.add(new ArrayList<>(currPosition));
            return;
        }
        for (int col = 0; col < n; col++) {
            currPosition.set(row, col);
            if (isValid(row, n, currPosition)) {
                recur(row + 1, n, currPosition, ans);
            }
        }
    }

    private static boolean isValid(int row, int n, List<Integer> currPosition) {
        int latestCol = currPosition.get(row);
        for (int i = 0; i < row; i++) {
            if (currPosition.get(i) == latestCol) return false;
            if (currPosition.get(i) - i == latestCol - row) return false;
            if (currPosition.get(i) + i == latestCol + row) return false;
        }
        return true;
    }

    @EpiTestComparator
    public static boolean comp(List<List<Integer>> expected,
                               List<List<Integer>> result) {
        if (result == null) {
            return false;
        }
        expected.sort(new LexicographicalListComparator<>());
        result.sort(new LexicographicalListComparator<>());
        return expected.equals(result);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "NQueens.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/n_queens.cc
#include <algorithm>
#include <iterator>
#include <vector>

#include "test_framework/generic_test.h"
using namespace std;

bool ok(int col, const vector<int> &curr_arrangement) {
    for (int row = 0; row < curr_arrangement.size(); row++) {
        if (curr_arrangement[row] == col) return false;
        if (abs(col - curr_arrangement[row]) == curr_arrangement.size() - row) return false;
    }
    return true;
}

void recur(int r, int n, vector<int> &curr_arrangement, vector<vector<int>> &ans) {
    if (r == n) {
        const auto successful_arrangement = curr_arrangement;
        ans.push_back(successful_arrangement);
        return;
    }
    for (int col = 0; col < n; col++) {
        if (ok(col, curr_arrangement)) {
            curr_arrangement.push_back(col);
            recur(r + 1, n, curr_arrangement, ans);
            curr_arrangement.pop_back();
        }
    }
}

vector<vector<int> > NQueens(int n) {
    vector<vector<int>> ans;
    vector<int> curr_arrangement;
    recur(0, n, curr_arrangement, ans);
    return ans;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"n"};
    return GenericTestMain(args, "n_queens.cc", "n_queens.tsv", &NQueens,
                           UnorderedComparator{}, param_names);
}
